lib/machine_gun/worker.ex

```diff
@@ -279,10 +282,10 @@ defmodule MachineGun.Worker do
          stream_ref,
          is_fin,
          from,
-         response,
+         %Response{headers: headers} = response,
          cancel_ref
        ) do
-    if is_fin == :fin do
+    if is_fin == :fin && !is_last_request(headers) do
       :ok = GenServer.reply(from, {:ok, response})
```

Comment:
What does this mean in practice? We don't reply with `:ok` if the response contains connection close header?

Reply:
yes, if the response contains connection close header we will reply only after the connection is closed (line 261)

Reply:
Do we expect server initiate the closing after it returned connection close header?

Reply:
no, connection close header means that server wants to close the connection and client shouldn't be sending more requests on it. I traced packets with Wireshark and `FIN,ACK` frame is sent by a client side (gun) after receiving this header. after connection is down we receive :gun_down and :DOWN messages and the latter one resets `gun_pid` to nil. next time the worker is invoked new connection is gonna be created.

Reply:
What if we only make this change:

```
if is_fin == :fin do
  :ok = GenServer.reply(from, {:ok, response})
  worker = clean_refs(worker, stream_ref, cancel_ref)
  if last_request?(headers) do
    :ok = :gun.close(gun_pid)      
    true = :erlang.demonitor(gun_ref, [:flush])
    %{worker | gun_pid: nil, gun_ref: nil}
  else
    worker
  end
  worker
end

Reply:
yes, it's much simpler. I've made suggested changes plus added clearing `streams` and `cancels` from the state after connection is terminated.